README.markdown
# [Django Macros URL](https://github.com/phpdude/django-macros-url/) v0.4.0 - Routing must be simple as possible

Django Macros URL makes it easy to write (and read) URL patterns in your Django applications by using macros.

You can combine your prefixes with macro names with an underscore, for example, you can use a macro `:slug` 
and `:product_slug`. They both will be compiled to same regex pattern with their group names of course. 
Multiple underscores accepted too.

[![Build Status](https://travis-ci.org/phpdude/django-macros-url.svg?branch=master)](https://travis-ci.org/phpdude/django-macros-url)

### Supported macros by default

```
slug - [\w-]+
year - \d{4}
month - (0?([1-9])|10|11|12)
day - ((0|1|2)?([1-9])|[1-3]0|31)
id - \d+
pk - \d+
page - \d+
uuid - [a-fA-F0-9]{8}-?[a-fA-F0-9]{4}-?[1345][a-fA-F0-9]{3}-?[a-fA-F0-9]{4}-?[a-fA-F0-9]{12}
```

If you want to offer more macros by default, you can fork and make a pull request.

### Installation

You can install the library with PyPI.

```
pip install django-macros-url
```

### Usage

Django Macros URLs used the same way as Django standard URLs. You just import this and declare your 
patterns with macros.

Also, you can register new macro (or maybe you want to replace default macro with your like regex
pattern) with `macrosurl.register(macro, pattern)` method.

An example of registration.

```python
import macrosurl

macrosurl.register('myhash', '[a-f0-9]{9}')

urlpatterns = patterns(
    'yourapp.views',
    macrosurl.url('^:myhash/$', 'myhash_main'),
    macrosurl.url('^news/:news_myhash/$', 'myhash_news'),
)
```

Feel free to register custom macro anywhere (i do it in main urls.py file). Macros URLs uses lazy 
initialization. Macros will be compiled only on the first request.

### URL normalization

Once Macros URL completed compile regex pattern, it makes normalization of it by rules:

- Strip from left side all whitespace and ^
- Strip from right side of pattern all whitespace and $
- Add to left side ^
- Add to right side $

This makes your URLs always very strong to adding any unexpected params into a path.

### Auto-calling as_view() on CBV objects.

Library check type of view and if a view is type object with defined 'as_view' function, call this. This allows 
you omit ".as_view()" calls in your urls.py files. But you can call this manual with params if you need.

This feature helps you to keep your urls.py files clean as possible. I hope you like this feature!

### Examples

Macros URL example urls.py file

```python
from django.conf.urls import patterns
from macrosurl import url
from project.portal.views import IndexView

urlpatterns = patterns(
    'yourapp.views',
    url('^:category_slug/$', 'category'),
    url(':category_slug/:product_slug/', 'category_product'),
    url(':category_slug/:product_slug/:variant_id', 'category_product_variant'),
    url('news/', 'news'),
    url('news/:year/:month/:day', 'news_date'),
    url('news/:slug', 'news_entry'),
    url('^order/:id$', 'order'),
    url('^$', IndexView),
)
```

Standard Django urls example

```python
from django.conf.urls import patterns, url
from project.portal.views import IndexView


urlpatterns = patterns(
    'yourapp.views',
    url('^(?P<category_slug>[\w-]+>)/$', 'category'),
    url('^(?P<category_slug>[\w-]+>)/(?P<product_slug>[\w-]+>)/$', 'category_product'),
    url('^(?P<category_slug>[\w-]+>)/(?P<product_slug>[\w-]+>)/(?P<variant_id>\d+>)$', 'category_product_variant'),
    url('^news/$', 'news'),
    url('^news/(?P<year>\d{4}>)/(?P<month>(0?([1-9])|10|11|12)>)/(?P<day>((0|1|2)?([1-9])|[1-3]0|31)>)$', 'news_date'),
    url('^news/(?P<slug>[\w-]+>)$', 'news_entry'),
    url('^order/(?P<id>\d+>)$', 'order'),
    url('^$', IndexView.as_view()),
)
```

I think you understand the difference of ways :)

#### Routing must be simple! ;-)

I think raw URL regexp patterns needed in 1% case only. I prefer simple way to write (and read, this is 
important) fancy clean URLs.

### Contributor

[Alexandr Shurigin](https://github.com/phpdude/)

You are welcome to contribute by PR.

CHANGELOG.markdown
v0.4.0
------

Added Py3.{5,6} & Django 1.11. Removed Py2.6 & Django <1.8 (#9)
* Quick fix for the DJANGO_VERSION error when using Python >= 3.6 (version calculation caching removed because it is not a performance penalty, urls are lazy and are calculated only once per an app start).

* Added Py3.{5,6} & Django 1.11. Removed Py2.6 & Django <1.8

The latest Django LTS is 1.11. Its ok to support up to the "LTS -1" (1.8) only.
The next Django LTS is 2.0, and will be Py3k only, so lets start testing with Py3k!

* Mark as Py3k compat on setup.py

v0.3.1
------

Fix for https://github.com/phpdude/django-macros-url/issues/8.

v0.3.0
------

Added support for Django 1.10 version.

v0.2.3
------

Development Status changed to "Development Status :: 5 - Production/Stable".

v0.2.2
------

Reverting tests back.

v0.2.1
------

Added `django.setup()` for tests, because looks like now it is required. README.markdown fixes.

v0.2.0
------

Added auto-calling as_view on CBVs objects. Now you can omit as_view() in your views by default. 
Have your code more clean then before ;-)

v0.1.7
------

setup.py fixes in requires.

v0.1.6
------

Tests updated to support Django1.8. Travis configuration updated too. Small cleanups.

v0.1.5
------

Add uuid version to uuid regex

v0.1.4
------

Readme fixes.

v0.1.3
------

macrosurl.__init__ django import fix to allow install macros-url when django is not installed yet.

v0.1.2
------

added pk macros

v0.1.1
------

Support for include('path.to.url') view added. Urls normalization by adding dollar at end fixed to support include. Tests added.

v0.1.0
------

Basic functionality done.

tests/urls.py
import os
import warnings

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "tests.settings")

import sys
import uuid

from django.conf.urls import include

from macrosurl import MacroUrlPattern, url

if sys.version_info >= (2, 7):
    import unittest
else:
    from django.utils import unittest


class TestRegexCompilation(unittest.TestCase):
    def setUp(self):
        warnings.filterwarnings("ignore")

    def test_nomacro(self):
        self.assertEqual(MacroUrlPattern('^$').compiled, '^$')
        self.assertEqual(MacroUrlPattern('^news/all/$').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/all/$').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/all/$').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/:news/$').compiled, '^news/:news/$')

    def test_normalize_url(self):
        self.assertEqual(MacroUrlPattern('').compiled, '^$')
        self.assertEqual(MacroUrlPattern('news/all/').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/all/$').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/all/').compiled, '^news/all/$')

    def test_strip_whitespace(self):
        self.assertEqual(MacroUrlPattern('').compiled, '^$')
        self.assertEqual(MacroUrlPattern(' news/all/').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern('^news/all/$ ').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern(' ^news/all/ ').compiled, '^news/all/$')
        self.assertEqual(MacroUrlPattern(' ^news/all/ \n').compiled, '^news/all/$')

    def test_id(self):
        self.assertEqual(MacroUrlPattern('page/:id').compiled, '^page/(?P<id>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:product_id').compiled, '^product/(?P<product_id>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:id/:product_id').compiled,
                         '^product/(?P<id>\d+)/(?P<product_id>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:id/:product_id/:news_id').compiled,
                         '^product/(?P<id>\d+)/(?P<product_id>\d+)/(?P<news_id>\d+)$')

    def test_pk(self):
        self.assertEqual(MacroUrlPattern('page/:pk').compiled, '^page/(?P<pk>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:product_pk').compiled, '^product/(?P<product_pk>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:pk/:product_pk').compiled,
                         '^product/(?P<pk>\d+)/(?P<product_pk>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:pk/:product_pk/:news_pk').compiled,
                         '^product/(?P<pk>\d+)/(?P<product_pk>\d+)/(?P<news_pk>\d+)$')

    def test_page(self):
        self.assertEqual(MacroUrlPattern('page/:page').compiled, '^page/(?P<page>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:product_page').compiled, '^product/(?P<product_page>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:page/:product_page').compiled,
                         '^product/(?P<page>\d+)/(?P<product_page>\d+)$')
        self.assertEqual(MacroUrlPattern('product/:page/:product_page/:news_page').compiled,
                         '^product/(?P<page>\d+)/(?P<product_page>\d+)/(?P<news_page>\d+)$')

    def test_slug(self):
        self.assertEqual(MacroUrlPattern('page/:slug').compiled, '^page/(?P<slug>[\w-]+)$')
        self.assertEqual(MacroUrlPattern('page/:category_slug/:slug').compiled,
                         '^page/(?P<category_slug>[\w-]+)/(?P<slug>[\w-]+)$')

    def test_year(self):
        self.assertEqual(MacroUrlPattern('news/:year').compiled, '^news/(?P<year>\d{4})$')

    def test_year_month(self):
        self.assertEqual(MacroUrlPattern('news/:year/:month').compiled,
                         '^news/(?P<year>\d{4})/(?P<month>(0?([1-9])|10|11|12))$')

    def test_year_month_day(self):
        self.assertEqual(MacroUrlPattern('news/:year/:month/:day/').compiled,
                         '^news/(?P<year>\d{4})/(?P<month>(0?([1-9])|10|11|12))/(?P<day>((0|1|2)?([1-9])|[1-3]0|31))/$')

    def test_date(self):
        self.assertEqual(MacroUrlPattern('news/:date/').compiled,
                         '^news/(?P<date>\d{4}-(0?([1-9])|10|11|12)-((0|1|2)?([1-9])|[1-3]0|31))/$')

    def test_uid(self):
        self.assertEqual(MacroUrlPattern('invoice/:uuid').compiled,
                         '^invoice/(?P<uuid>[a-fA-F0-9]{8}-?[a-fA-F0-9]{4}-?[1345][a-fA-F0-9]{3}-?[a-fA-F0-9]{4}-?['
                         'a-fA-F0-9]{12})$')

    def test_strongurl(self):
        self.assertEqual(MacroUrlPattern('orders/:date/:uuid/products/:slug/:variant_id').compiled,
                         '^orders/(?P<date>\\d{4}-(0?([1-9])|10|11|12)-((0|1|2)?([1-9])|[1-3]0|31))/(?P<uuid>['
                         'a-fA-F0-9]{8}-?[a-fA-F0-9]{4}-?[1345][a-fA-F0-9]{3}-?[a-fA-F0-9]{4}-?[a-fA-F0-9]{'
                         '12})/products/(?P<slug>[\\w-]+)/(?P<variant_id>\\d+)$')

    # noinspection PyProtectedMember
    def test_includes_end(self):
        self.assertEqual(str(url('users/:slug', include('tests'))._regex), '^users/(?P<slug>[\\w-]+)')
        self.assertEqual(str(url('users/:slug', include('tests', namespace='1'))._regex), '^users/(?P<slug>[\\w-]+)')
        self.assertEqual(str(url('users/:slug', 'tests.views.view')._regex), '^users/(?P<slug>[\\w-]+)$')


class TestRegexUrlResolving(unittest.TestCase):
    def setUp(self):
        self.view = 'tests.views.view'

        warnings.filterwarnings("ignore")

    def test_id(self):
        self.assertIsNone(url('product/:id', self.view).resolve('product/test'))
        self.assertIsNotNone(url('product/:id', self.view).resolve('product/10'))
        self.assertEqual(url('product/:id', self.view).resolve('product/10').kwargs['id'], '10')
        self.assertEqual(url('product/:product_id', self.view).resolve('product/10').kwargs['product_id'],
                         '10')

    def test_slug(self):
        self.assertIsNone(url('product/:slug', self.view).resolve('product/test/ouch'))
        self.assertIsNotNone(url('product/:slug', self.view).resolve('product/test'))
        self.assertIsNotNone(url('product/:slug/:other_slug', self.view).resolve('product/test/other'))

    def test_year(self):
        self.assertIsNone(url('news/:year', self.view).resolve('news/last'))
        for y in range(1970, 2025):
            self.assertIsNotNone(url('news/:year', self.view).resolve('news/%s' % y))
        self.assertIsNone(url('news/:year/last', self.view).resolve('news/2014/other'))
        self.assertIsNotNone(url('news/:year/last', self.view).resolve('news/2014/last'))

    def test_year_month(self):
        self.assertIsNone(url('news/:year/:month', self.view).resolve('news/2014/last'))
        self.assertIsNone(url('news/:year/:month', self.view).resolve('news/2014/2012'))
        for y in range(1970, 2025):
            for m in range(1, 12):
                self.assertIsNotNone(url('news/:year/:month', self.view).resolve('news/%s/%s' % (y, m)))

        self.assertIsNotNone(url('news/:year/:month/last', self.view).resolve('news/2014/12/last'))

    def test_year_month_day(self):
        self.assertIsNone(url('news/:year/:month/:day', self.view).resolve('news/2014/12/last'))
        self.assertIsNone(url('news/:year/:month/:day', self.view).resolve('news/2014/2012/31'))
        for y in range(2000, 2020):
            for m in range(1, 12):
                for d in range(1, 31):
                    self.assertIsNotNone(url('news/:year/:month/:day', self.view).resolve('news/%s/%s/%s' % (y, m, d)))

    def test_date(self):
        self.assertIsNone(url('news/:date', self.view).resolve('news/2014/12/12'))
        for y in range(2000, 2020):
            for m in range(1, 12):
                for d in range(1, 31):
                    self.assertIsNotNone(url('news/:date', self.view).resolve('news/%s-%s-%s' % (y, m, d)))

    def test_uuid(self):
        self.assertIsNone(url("invoice/:uuid", self.view).resolve('invoice/123123-123123-1231231-1231312-3-1312312-'))
        for i in range(1, 1000):
            self.assertIsNotNone(url("invoice/:uuid", self.view).resolve('invoice/%s' % uuid.uuid4()))

    def test_no_match_for_invalid_uuid(self):
        """
        UUID with invalid version.  The allowed versions are 1, 2, 4 and 5
        xxxxxxxx-xxxx-Vxxx-xxx-xxxxxxxxxxxx

        https://github.com/phpdude/django-macros-url/pull/2
        """
        self.assertIsNone(url("invoice/:uuid", self.view).resolve('invoice/3e41b04d-0978-9027-86c2-aa90c63ecb54'))

    def test_cdv_as_view_calling(self):
        from .views import CBVView

        self.assertIsInstance(url("", CBVView).resolve('').func, type(lambda: 1))

    def test_cdv_as_view_pass_manual(self):
        from .views import CBVView

        self.assertIsInstance(url("", CBVView.as_view()).resolve('').func, type(lambda: 1))

    def test_cdv_as_view_pass_manual_params(self):
        from .views import CBVView

        self.assertIsInstance(url("", CBVView.as_view(x='test.html')).resolve('').func, type(lambda: 1))


tests/views.py
from django.views.generic import View


def view(request):
    pass


class CBVView(View):
    x = None


tests/__init__.py
__author__ = 'dude'


macrosurl/__init__.py
import re
import warnings
from distutils.version import StrictVersion

VERSION = (0, 4, 0)

_macros_library = {
    'id': r'\d+',
    'pk': r'\d+',
    'page': r'\d+',
    'slug': r'[\w-]+',
    'year': r'\d{4}',
    'month': r'(0?([1-9])|10|11|12)',
    'day': r'((0|1|2)?([1-9])|[1-3]0|31)',
    'date': r'\d{4}-(0?([1-9])|10|11|12)-((0|1|2)?([1-9])|[1-3]0|31)',
    'uuid': r'[a-fA-F0-9]{8}-?[a-fA-F0-9]{4}-?[1345][a-fA-F0-9]{3}-?[a-fA-F0-9]{4}-?[a-fA-F0-9]{12}',
}


def get_version(*args, **kwargs):
    return ".".join(map(str, VERSION))


def register(macros, pattern):
    _macros_library[macros] = pattern


def regex_group(macro, pattern):
    return '(?P<%s>%s)' % (macro, pattern)


def normalize_pattern(url, end_dollar=True):
    pattern = '^%s$'
    if not end_dollar:
        pattern = '^%s'

    return pattern % url.lstrip("^ \n").rstrip("$ \n")


class MacroUrlPattern(object):
    def __init__(self, pattern, end_dollar=True):
        self.pattern = pattern
        self.end_dollar = end_dollar

    def compile(self):
        pattern = self.pattern
        macros = re.findall('(:([a-z_\d]+))', pattern)
        for match, macro in macros:
            if macro in _macros_library:
                pattern = pattern.replace(match, regex_group(macro, _macros_library[macro]))
            else:
                for _macro in _macros_library:
                    if macro.endswith("_%s" % _macro):
                        pattern = pattern.replace(match, regex_group(macro, _macros_library[_macro]))
                        continue

        return normalize_pattern(pattern, self.end_dollar)

    @property
    def compiled(self):
        if not hasattr(self, '_compiled'):
            setattr(self, '_compiled', self.compile())

        return getattr(self, '_compiled')

    def __str__(self):
        return self.compiled

    def __unicode__(self):
        return self.__str__()


def url(regex, view, kwargs=None, name=None, prefix=''):
    from django.conf.urls import url as baseurl
    from django import get_version

    # Handle include()'s in views.
    end_dollar = True
    if isinstance(view, tuple) and len(view) == 3:
        end_dollar = False

    # Auto-calling as_view on CBVs objects. Now you can omit as_view() in your views by default.
    if isinstance(view, type):
        if hasattr(view, 'as_view') and hasattr(view.as_view, '__call__'):
            view = view.as_view()

    if prefix:
        warnings.warn(
            'Support for prefix in macrosurl.url() is deprecated and '
            'will be removed in version 0.4 (support for prefix was removed in Django 1.10). '
            'Please update your source code.'
            'In old Django versions prefix was used like "if prefix:view = prefix + \'.\' + view".'
        )

        view = prefix + '.' + view

    if get_version() >= StrictVersion('1.10') and not callable(view) and not isinstance(view, (list, tuple)):
        warnings.warn(
            'View "%s" must be a callable in case of Django 1.10. '
            'Macrosurl will try to load the view automatically (this behavior will be removed in version 0.4), but '
            'please update your source code.' % view
        )

        from django.utils.module_loading import import_string

        view = import_string(view)

    return baseurl(MacroUrlPattern(regex, end_dollar=end_dollar), view, kwargs=kwargs, name=name)
